docs/tutorial_history_backend.rst

```diff
@@ -0,0 +1,319 @@
+        ts = int(time.time() * 1000)
+        return '{}-{}-{}'.format(self.sessionid, ts, str(uuid.uuid4())[:18])
+
+    def _request_db_data(self, path, data=None):
+        url = 'http://127.0.0.1:5984' + path
+        headers = {'Content-Type': 'application/json'}
+        if data is not None:
+            resp = requests.post(url, json.dumps(data), headers=headers)
+        else:
+            headers = {'Content-Type': 'text/plain'}
+            resp = requests.get(url, headers=headers)
+        return resp
+
+``_save_to_db()`` takes a dict as the input, which contains the information
+about a command that use input, and save it into CouchDB.
+
+Instead of letting CouchDB provide us a random Document ID (i.e. the
+``data['_id']`` in our code), we built it for ourselves.  We use the Unix
+timestamp and UUID string for a second time. Prefixing with ``self.sessionid``
+we have, we make history items in order inside a single xonsh session too.
+So that we don't need any extra CouchDB's
+`Design Documents and Views <http://docs.couchdb.org/en/2.0.0/couchapp/ddocs.html>`_
+feature. Just with a bare ``_all_docs`` API, we can fetch history items back
+in order.
+
+Now that we have helper functions, we can update our ``append()`` method
+to do the real job - save history into DB.
+
+.. code-block:: python
+
+    def append(self, cmd):
+        self.inps.append(cmd['inp'])
+        self.rtns.append(cmd['rtn'])
+        self.outs.append(None)
+        self.tss.append(cmd.get('ts', (None, None)))
+        self._save_to_db(cmd)
+
+This method will be called by xonsh every time it run a new command from user.
+
+
+Retrieve History Items
+======================
+
+.. code-block:: python
+
+    def items(self):
+        yield from self._get_db_items(self.sessionid)
+
+    def all_items(self):
+        yield from self._get_db_items()
+
+These two methods are responsible for get history items for current xonsh
+session and all historical sessions respectively.
+
+And here is our helper methods to get docs from DB:
+
+.. code-block:: python
+
+    def _get_db_items(self, sessionid=None):
+        path = '/xonsh-history/_all_docs?include_docs=true'
+        if sessionid is not None:
+            path += '&start_key="{0}"&end_key="{0}-z"'.format(sessionid)
+        try:
+            r = self._request_db_data(path)
+        except Exception as e:
+            msg = 'error when query db: {}: {}'.format(e.__class__.__name__, e)
+            print(msg, file=sys.stderr)
+            return
+        data = json.loads(r.text)
+        for item in data['rows']:
+            cmd = item['doc'].copy()
+            cmd['ts'] = cmd['ts'][0]
+            yield cmd
+
+The `try-except` is here so that we're safe when something bad happened, like
+couchdb is not get started, etc.
+
+
+Try Out Our New History Backend
+===============================
+
+That's it. Your can find full code here:
+`<https://gist.github.com/mitnk/2d08dc60aab33d8b8b758c544b37d570>`_
```

Comment:
Do we want to host this within the documentation? Just curious, I'm fine with it as a gist, too.